File: 2020/6/이중 우선 순위 큐.py
def solution(operations):
    hq = []
    for oper in operations:
        if oper[0] == "I":
            hq.append(int(oper.split(" ")[1]))
        elif len(hq):
            if oper == "D -1":
                hq.remove(min(hq))
            elif oper == "D 1":
                hq.remove(max(hq))
    if len(hq):
        return [ max(hq), min(hq)]
    else:
        return  [0,0]


import heapq
def solution(operations):
    hq = []
    for oper in operations:
        if oper[0] == "I":
            heapq.heappush(hq,int(oper.split(" ")[1]))
        elif len(hq):
            if oper == "D -1":
                heapq.heappop(hq)
            elif oper == "D 1":
                hq.remove(max(hq))
    if len(hq):
        return [ max(hq), heapq.heappop(hq)]
    else:
        return  [0,0]

File: 2020/11/이중 우선 순위 큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = {};
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(String operation : operations){
            String []value = operation.split(" ");
            if(value[0].equals("I")){
                pq.offer(Integer.parseInt(value[1]));
            }else if(value[0].equals("D") &&  value[1].equals("1")){
                if(!pq.isEmpty()){
                    int max = Collections.max(pq);
                    pq.remove(Integer.valueOf(max));
                }
            }else if(value[0].equals("D") &&  value[1].equals("-1")){
                if(!pq.isEmpty()){
                    pq.poll();
                }
            }
        }
        return pq.isEmpty() ? new int[]{0, 0}:  new int[]{Collections.max(pq), pq.poll()};
    }
}